### PR TITLE
finished chapter 3

### PR DESCRIPTION
**Pull request by keuhdall**
### Solutions for Chapter 3
(also did some changes in chapter 2 exercices)

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -892,9 +894,10 @@ and reverses it.
   cheating!
 -}
 rewind :: [a] -> [a]
-rewind [] = []
-rewind [x] = [x]
-rewind (x:xs) = rewind xs ++ [x]
+rewind xs = go xs [] where
```
Comment (keuhdall) [inline src/Chapter2.hs:897]: Not sure if I understood correctly the go pattern, is that how I was supposed to do it ?

File: src/Chapter2.hs [hunk 2]
```diff
@@ -892,9 +894,10 @@ and reverses it.
...
+  go :: [a] -> [a] -> [a]
+  go [] acc = acc
+  go (y:ys) acc = go ys (y:acc)
```
Comment (keuhdall) [inline src/Chapter2.hs:900]: Not sure if I understood correctly the go pattern, is that how I was supposed to do it ?

Comment (chshersh) [reply src/Chapter2.hs:900]: Yes, that's an absolutely correct implementation :ok_hand: You nailed it :hammer:

File: src/Chapter3.hs [hunk 3]
```diff
@@ -344,6 +346,15 @@ of a book, but you are not limited only by the book properties we described.
 Create your own book type of your dreams!
 -}
 
+data Genre = Fantasy | Novel | Thriller | SciFi | Other(String) deriving (Eq, Show)
```
Comment (chshersh) [inline src/Chapter3.hs]: Wow, that's nice work on data types here! Very type-safe modelling :+1: 
I have only one minor stylistic suggestion. Usually, types are not take in `()` if they don't have type parameters 
```suggestion
data Genre = Fantasy | Novel | Thriller | SciFi | Other String
    deriving (Eq, Show)
```

File: src/Chapter3.hs [hunk 4]
```diff
@@ -374,6 +385,26 @@ after the fight. The battle has the following possible outcomes:
...
+} deriving (Eq, Show)
+
+data Monster = Monster {
+  monsterHp :: Int,
+  monsterAttack :: Int,
+  monsterGold :: Int
+} deriving (Eq, Show)
+
+fight :: Knight -> Monster -> Int
+fight Knight{..} Monster{..}
```
Comment (chshersh) [inline src/Chapter3.hs:402]: That's a very nice usage of RecordWildCards :tiger: 

Shameless plug, if you want to read more about best-practices around RecordWildCards, you can read my blog post on the subject :book: 

* https://kodimensional.dev/recordwildcards

File: src/Chapter3.hs [hunk 5]
```diff
@@ -460,6 +491,8 @@ Create a simple enumeration for the meal types (e.g. breakfast). The one who
 comes up with the most number of names wins the challenge. Use your creativity!
 -}
 
+data MealTypes = Breakfast | Brunch | Lunch | TeaTime | Dinner | Snack | AfterPartyMeal
```
Comment (chshersh) [inline src/Chapter3.hs:494]: Nice menu! Though, the party probably wasn't that great, if you still need a meal after that party :unamused: :rofl:

File: src/Chapter3.hs [hunk 6]
```diff
@@ -480,6 +513,42 @@ After defining the city, implement the following functions:
...
+} deriving Show
+
+data City = City {
+  castle :: Castle,
+  building :: Building,
+  houses :: [House]
+} deriving Show
+
+buildCastle :: City -> String -> City
+buildCastle city castleName = city{castle = Castle(castleName)}
```
Comment (chshersh) [inline src/Chapter3.hs]: Constructors in Haskell are ordinary functions. They just start with an uppercase letter :top: 
So usual function calling conventions are applied here.

Also, I see that if the `Castle` already had walls, they will be removed by this function :bricks: :european_castle: :bricks: 
```suggestion
buildCastle city castleName = city{castle = Castle castleName}
```

File: src/Chapter3.hs [hunk 7]
```diff
@@ -480,6 +513,42 @@ After defining the city, implement the following functions:
...
+buildHouse :: City -> House -> City
+buildHouse city@City{houses} house = city{houses = house:houses}
+
+
+buildWalls :: City -> City
+buildWalls city@City{castle,houses} = case castle of
+  Castle castleName -> if pplCount > 10 then city{castle = CastleWithWalls(castleName)} else city
+  _ -> city
+  where
+    pplCount = foldl (\acc house' -> acc + (pplInHouse . peopleAmount $ house')) (0 :: Int) houses
```
Comment (chshersh) [inline src/Chapter3.hs]: Nice usage of standard functions! If you want to use `foldl`, then it's always better to use `foldl'` from the `Data.List` module. Ordinary `foldl` leaks memory, unfortunately :potable_water: 

Alternatively, you can use a combination of `map` and `sum` functions from `Prelude`. But your solution is already correct, these are just some notes :slightly_smiling_face:

File: src/Chapter3.hs [hunk 8]
```diff
@@ -560,22 +629,37 @@ introducing extra newtypes.
...
+    , playerArmor     :: Armor
+    , playerAttack    :: Attack
+    , playerDexterity :: Dexterity
+    , playerStrength  :: Strength
     }
 
-calculatePlayerDamage :: Int -> Int -> Int
-calculatePlayerDamage attack strength = attack + strength
+calculatePlayerDamage :: Attack -> Strength -> Damage
+calculatePlayerDamage (Attack a) (Strength s) = Damage $ a + s
```
Comment (chshersh) [inline src/Chapter3.hs:658]: Nice work on newtypes! :medal_sports: Newtypes are one very important feature in Haskell :+1:
You can read more about the newtype pattern in our blog post: https://kowainik.github.io/posts/haskell-mini-patterns#newtype

File: src/Chapter3.hs [hunk 9]
```diff
@@ -971,6 +1080,23 @@ implement the following functions:
...
+                | Sunday deriving (Eq, Bounded, Enum)
+
+isWeekend :: DaysOfWeek -> Bool
+isWeekend day = day == Saturday || day == Sunday
+
+nextDay :: DaysOfWeek -> DaysOfWeek
+nextDay day = if day == maxBound then minBound else succ day
+
+daysToParty :: DaysOfWeek -> Int
+daysToParty day = (fromEnum Friday - fromEnum day) `mod` 7
```
Comment (chshersh) [inline src/Chapter3.hs:1100]: Very math solution! :1234:

**Review by chshersh — APPROVED** [review]
Great work! Very creative and nice solutions :art: 

I left some comments, some suggestions and possible ideas for improvements. But your solutions are already great :+1:

File: src/Chapter3.hs (hunk 4 above)
Comment (keuhdall) [reply src/Chapter3.hs:402]: Will definitely take a look, thank you ! 😄

**Pull request merged**